File: Week_01/MyCircularDeque.c

#include <stdio.h>
#include <stdlib.h>
#include <stdbool.h>

typedef struct {
    int *arr;
	int head;
	int tail;
	int size;
} MyCircularDeque;

bool myCircularDequeIsFull(MyCircularDeque* obj);
bool myCircularDequeIsEmpty(MyCircularDeque* obj);

/** Initialize your data structure here. Set the size of the deque to be k. */

MyCircularDeque* myCircularDequeCreate(int k) {
    MyCircularDeque *obj = (MyCircularDeque *)malloc(sizeof(MyCircularDeque));
    obj->head = 0;
    obj->tail = 0;
    obj->size = k + 1;
    obj->arr = (int *)malloc(sizeof(int) * (k + 1));
    return obj;
}

/** Adds an item at the front of Deque. Return true if the operation is successful. */
bool myCircularDequeInsertFront(MyCircularDeque* obj, int value) {
    if (myCircularDequeIsFull(obj)) return false;
	int pos = (obj->head + obj->size - 1) % obj->size;
	obj->arr[pos] = value;
	obj->head = pos;
	return true;
}

/** Adds an item at the rear of Deque. Return true if the operation is successful. */
bool myCircularDequeInsertLast(MyCircularDeque* obj, int value) {
    if (myCircularDequeIsFull(obj)) return false;
    obj->arr[obj->tail] = value;
    obj->tail = (obj->tail + 1) % obj->size;
    return true;
}

/** Deletes an item from the front of Deque. Return true if the operation is successful. */
bool myCircularDequeDeleteFront(MyCircularDeque* obj) {
    if (myCircularDequeIsEmpty(obj)) return false;
    obj->head = (obj->head + 1) % obj->size;
    return true;
}

/** Deletes an item from the rear of Deque. Return true if the operation is successful. */
bool myCircularDequeDeleteLast(MyCircularDeque* obj) {
    if (myCircularDequeIsEmpty(obj)) return false;
    obj->tail = (obj->size + obj->tail - 1) % obj->size;
    return true;
}

/** Get the front item from the deque. */
int myCircularDequeGetFront(MyCircularDeque* obj) {
    if (myCircularDequeIsEmpty(obj)) return -1;
    return obj->arr[obj->head];
}

/** Get the last item from the deque. */
int myCircularDequeGetRear(MyCircularDeque* obj) {
    if (myCircularDequeIsEmpty(obj)) return -1;
    int pos = (obj->tail - 1 + obj->size) % obj->size;
    return obj->arr[pos];
}

/** Checks whether the circular deque is empty or not. */
bool myCircularDequeIsEmpty(MyCircularDeque* obj) {
    return (obj->head == obj->tail) ? true : false;
}

/** Checks whether the circular deque is full or not. */
bool myCircularDequeIsFull(MyCircularDeque* obj) {
    return ((obj->head - 1 + obj->size) % obj->size == obj->tail) ? true : false;
}

int myCircularDequeLength(MyCircularDeque* obj) {
    if (obj->tail > obj->head) {
        return obj->tail - obj->head;
    } else {
        return obj->tail - obj->head + obj->size;
    }
}

void myCircularDequeFree(MyCircularDeque* obj) {
    free(obj->arr);
	free(obj);
	return;
}

void PrintDeque(MyCircularDeque *obj);

int main(int argc, const char *argv[]) {
    MyCircularDeque *obj = myCircularDequeCreate(5);
    myCircularDequeInsertFront(obj, 10);
    myCircularDequeInsertFront(obj, 20);
    myCircularDequeInsertFront(obj, 30);
    myCircularDequeInsertLast(obj, 0);
    myCircularDequeInsertLast(obj, 1);
    myCircularDequeDeleteLast(obj);
    int front = myCircularDequeGetFront(obj);
    int rear = myCircularDequeGetRear(obj);
    printf("front=%d, rear=%d\n", front, rear);
    PrintDeque(obj);
    return 0;
}

// 打印
void PrintDeque(MyCircularDeque *obj) {
    printf("head=%d, tail=%d, size=%d\n", obj->head, obj->tail, obj->size);
    printf("arr=[");
    int length = myCircularDequeLength(obj);
    int max = obj->head + length;
    for (int i = obj->head; i < max; i++) {
        printf("%d", obj->arr[i % obj->size]);
        if (i != max - 1) {
            printf(", ");
        }
    }
    printf("]\n");
}

File: Week_03/MyCircularDeque.swift
//
//  MyCircularDeque.swift
//  LeetCodeDemo
//
//  Created by Yangdongwu on 2020/12/18.
//

import Foundation

class MyCircularDeque {
    var head: Int
    var tail: Int
    let size: Int
    var array: [Int]
    /** Initialize your data structure here. Set the size of the deque to be k. */
    init(_ k: Int) {
        head = 0
        tail = 0
        size = k + 1
        array = [Int](repeating: 0, count: k + 1)
    }
    
    /** Adds an item at the front of Deque. Return true if the operation is successful. */
    func insertFront(_ value: Int) -> Bool {
        if isFull() {
            return false
        }
        head = (head - 1 + size) % size
        array[head] = value
        return true
    }
    
    /** Adds an item at the rear of Deque. Return true if the operation is successful. */
    func insertLast(_ value: Int) -> Bool {
        if isFull() {
            return false
        }
        array[tail] = value
        tail = (tail + 1) % size
        return true
    }
    
    /** Deletes an item from the front of Deque. Return true if the operation is successful. */
    func deleteFront() -> Bool {
        if isEmpty() {
            return false
        }
        head = (head + 1) % size
        return true
    }
    
    /** Deletes an item from the rear of Deque. Return true if the operation is successful. */
    func deleteLast() -> Bool {
        if isEmpty() {
            return false
        }
        tail = (tail - 1 + size) % size
        return true
    }
    
    /** Get the front item from the deque. */
    func getFront() -> Int {
        if isEmpty() {
            return -1
        }
        return array[head]
    }
    
    /** Get the last item from the deque. */
    func getRear() -> Int {
        if isEmpty() {
            return -1
        }
        let pos = (tail - 1 + size) % size
        return array[pos]
    }
    
    /** Checks whether the circular deque is empty or not. */
    func isEmpty() -> Bool {
        return head == tail
    }
    
    /** Checks whether the circular deque is full or not. */
    func isFull() -> Bool {
        return head == ((tail + 1) % size)
    }
    
}

/**
 * Your MyCircularDeque object will be instantiated and called as such:
 * let obj = MyCircularDeque(k)
 * let ret_1: Bool = obj.insertFront(value)
 * let ret_2: Bool = obj.insertLast(value)
 * let ret_3: Bool = obj.deleteFront()
 * let ret_4: Bool = obj.deleteLast()
 * let ret_5: Int = obj.getFront()
 * let ret_6: Int = obj.getRear()
 * let ret_7: Bool = obj.isEmpty()
 * let ret_8: Bool = obj.isFull()
 */

extension MyCircularDeque {
    class func test() {
        let deque = MyCircularDeque(5)
        print(deque.getRear())
        print(deque.getFront())
        print(deque.insertLast(1))
        print(deque.insertFront(4))
        print(deque.insertLast(2))
        print(deque.insertLast(3))
        print(deque.insertLast(5))
        print(deque.insertLast(6))
        print(deque.head, deque.tail, deque.size)
        print(deque.array)
        print(deque.isFull())
        print(deque.isEmpty())
        print(deque.getRear())
        print(deque.getFront())
    }
}
